### PR TITLE
Add density current ( falling bubble example) with Laplacian diffusion

### LINKED ISSUE #577: Add density current test case
Check boundary conditions. Compare curl form and flux form solutions.

### PR DESCRIPTION
**Pull request by akshaysridhar**
Adds density current example (plane configuration). Constant coefficient, Laplacian viscosity applied as diffusion mechanism. Hyperdiffusion turned off in default configuration for these examples. 

Closes #577

### CODE REVIEW

File: examples/hybrid/density_current_2dinvariant_rhoe.jl [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+    cv_d = C_v
+    p_0 = MSLP
+    g = grav
+
+    # auxiliary quantities
+    r = sqrt((x - x_c)^2/x_r^2 + (z - z_c)^2/z_r^2)
+    θ_p = r < r_c ? 0.5 * θ_c * (1.0 + cospi(r / r_c)) : 0.0 # potential temperature perturbation
+
+    θ = θ_b + θ_p # potential temperature
+    π_exn = 1.0 - g * z / cp_d / θ # exner function
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhoe.jl:89]:
```suggestion
    π_exn = 1.0 - Φ(z) / cp_d / θ # exner function
```

File: examples/hybrid/density_current_2dinvariant_rhoe.jl [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    T = π_exn * θ # temperature
+    p = p_0 * π_exn^(cp_d / R_d) # pressure
+    ρ = p / R_d / T # density
+    e = cv_d * (T - T_0) + g * z
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhoe.jl:93]:
```suggestion
    e = cv_d * (T - T_0) + Φ(z)
```

File: examples/hybrid/density_current_2dinvariant_rhoe.jl [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    @. duₕ -= hgrad(cE)
+    @. dw -= vgradc2f(cE)
+
+    # 3) total energy
+
+    @. dρe -= hdiv(cuw * (cρe + cp))
+    @. dρe -= vdivf2c(fw * Ic2f(cρe + cp))
+    @. dρe -= vdivf2c(Ic2f(cuₕ * (cρe + cp)))
+    
+    # Vertical Order 2 Diffusion
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhoe.jl:228]: Is this both horizontal and vertical?

File: examples/hybrid/density_current_2dinvariant_rhoe.jl [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+
+    ### HYPERVISCOSITY
+    # 1) compute hyperviscosity coefficients
+    χe = @. dρe = hwdiv(hgrad(h_tot)) # we store χe in dρe
+    χuₕ = @. duₕ = hwgrad(hdiv(cuₕ))
+
+    Spaces.weighted_dss!(dρe)
+    Spaces.weighted_dss!(duₕ)
+
+    κ₄ = 98311.0 # m^4/s
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhoe.jl:160]: Do we still need hyperdiffusion if there is second order diffusion?

File: examples/hybrid/density_current_2dinvariant_rhotheta.jl [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    cv_d = C_v
+    p_0 = MSLP
+    g = grav
+
+    # auxiliary quantities
+    r = sqrt((x - x_c)^2/x_r^2 + (z - z_c)^2/z_r^2)
+    θ_p = r < r_c ? 0.5 * θ_c * (1.0 + cospi(r / r_c)) : 0.0 # potential temperature perturbation
+
+    θ = θ_b + θ_p # potential temperature
+    π_exn = 1.0 - g * z / cp_d / θ # exner function
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhotheta.jl:95]:
```suggestion
    π_exn = 1.0 - Φ(z) / cp_d / θ # exner function
```

File: examples/hybrid/density_current_2dinvariant_rhotheta.jl [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+    ### HYPERVISCOSITY
+    # 1) compute hyperviscosity coefficients
+
+    χθ = @. dρθ = hwdiv(hgrad(cρθ / cρ)) # we store χθ in dρθ
+    χuₕ = @. duₕ = hwgrad(hdiv(cuₕ))
+    Spaces.weighted_dss!(dρθ)
+    Spaces.weighted_dss!(duₕ)
+
+    κ₄ = 98311.0 # m^4/s
```
Comment (szy21) [inline examples/hybrid/density_current_2dinvariant_rhotheta.jl:150]: same as before, do we still need hyperdiffusion?

File: examples/hybrid/density_current_flux_form.jl [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    cv_d = C_v
+    p_0 = MSLP
+    g = grav
+
+    # auxiliary quantities
+    r = sqrt((x - x_c)^2/x_r^2 + (z - z_c)^2/z_r^2)
+    θ_p = r < r_c ? 0.5 * θ_c * (1.0 + cospi(r / r_c)) : 0.0 # potential temperature perturbation
+
+    θ = θ_b + θ_p # potential temperature
+    π_exn = 1.0 - g * z / cp_d / θ # exner function
```
Comment (szy21) [inline examples/hybrid/density_current_flux_form.jl:95]:
```suggestion
    π_exn = 1.0 - Φ(z) / cp_d / θ # exner function
```

File: examples/hybrid/density_current_flux_form.jl [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+    ### HYPERVISCOSITY
+    # 1) compute hyperviscosity coefficients
+    @. dρθ = hdiv(hgrad(θ))
+    @. dρuₕ = hdiv(hgrad(uₕ))
+    @. dρw = hdiv(hgrad(w))
+    Spaces.weighted_dss!(dYc)
+    Spaces.weighted_dss!(dρuₕ)
+    Spaces.weighted_dss!(dρw)
+
+    κ₄ = 100.0 # m^4/s
```
Comment (szy21) [inline examples/hybrid/density_current_flux_form.jl:224]: Same as before, do we still need hyperdiffusion? And any reason the hyperdiffusivity is different in the flux and advection forms?

File: examples/hybrid/density_current_2dinvariant_rhoe.jl (hunk 3 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_2dinvariant_rhoe.jl:228]: Yes; dangling comment from sequentially adding / testing vertical first, then horizontal. Will update.

File: examples/hybrid/density_current_flux_form.jl (hunk 8 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_flux_form.jl:224]: No - the coefficients have been modified to be consistent across all 3 examples ( results in shared drive reflect this)

File: examples/hybrid/density_current_2dinvariant_rhotheta.jl (hunk 6 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_2dinvariant_rhotheta.jl:150]: We do not necessarily need it; but following this : https://gmd.copernicus.org/articles/9/2007/2016/ I have included both 4th and 2nd order coeff in this case.

File: examples/hybrid/density_current_2dinvariant_rhoe.jl (hunk 4 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_2dinvariant_rhoe.jl:160]: No, we do not necessarily need this - see also the rho-theta example comment .

File: examples/hybrid/density_current_2dinvariant_rhotheta.jl (hunk 6 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_2dinvariant_rhotheta.jl:150]: Set \kappa\_4 == 0 for now ; in ClimaAtmos we will support both methods simultaneously

**Comment by akshaysridhar** [comment]
bors try

File: examples/hybrid/density_current_2dinvariant_rhoe.jl [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    @. dρe -= hdiv(cuw * (cρe + cp))
+    @. dρe -= vdivf2c(fw * Ic2f(cρe + cp))
+    @. dρe -= vdivf2c(Ic2f(cuₕ * (cρe + cp)))
+
+    # Uniform 2nd order diffusion
+    ∂c = Operators.GradientF2C()
+    fρ = @. Ic2f(cρ)
+    κ₂ = 75.0 # m^2/s
+
+    ᶠ∇ᵥuₕ = @. vgradc2f(cuₕ.components.data.:1)
```
Comment (jiahe23) [inline examples/hybrid/density_current_2dinvariant_rhoe.jl:233]: Does it work if we write it as `ᶠ∇ᵥuₕ = @. vgradc2f(cuₕ)`? If so, this formula generalizes to 3d cases as well.

Comment (jiahe23) [reply examples/hybrid/density_current_2dinvariant_rhoe.jl:233]: In the Ekman layer column case, (u, v) horizontal velocity are used and computed as https://github.com/CliMA/ClimaCore.jl/blob/main/examples/column/hydrostatic_ekman.jl#L148

**Comment by szy21** [comment]
Let's not worry about rho_eint and the unified driver for this PR. Could you check the above Jia's question on diffusion? Other than that it looks good to me, thanks!

**Review by szy21 — APPROVED** [review]

File: examples/hybrid/density_current_2dinvariant_rhoe.jl (hunk 9 above)
Comment (akshaysridhar) [reply examples/hybrid/density_current_2dinvariant_rhoe.jl]: Since I'm treating these variables as scalars, getting the components seems to be the right way to handle it here ; but we should be able to generalise this to vectors in the future.

**Comment by akshaysridhar** [comment]
bors try

File: .buildkite/pipeline.yml [hunk 10]
```diff
@@ -374,6 +374,39 @@ steps:
       config: cpu
       queue: central
       slurm_ntasks: 1
+  
+  - label: ":computer: Density current 2D hybrid invariant total energy"
+    key: "cpu_density_current_2d_hybrid_invariant_total_energy"
+    command:
+      - "julia --color=yes --project=examples examples/hybrid/plane/density_current_2dinvariant_rhoe.jl"
+    artifact_paths:
+      - "examples/hybrid/output/dc_invariant_etot/*"
```
Comment (szy21) [inline .buildkite/pipeline.yml:383]:
```suggestion
      - "examples/hybrid/output/plane/dc_invariant_etot/*"
```

File: .buildkite/pipeline.yml [hunk 11]
```diff
@@ -374,6 +374,39 @@ steps:
...
+      config: cpu
+      queue: central
+      slurm_ntasks: 1
+  
+  - label: ":computer: Density current 2D hybrid conservative form potential temperature"
+    key: "cpu_density_current_2d_hybrid_conservative_potential_temperature"
+    command:
+      - "julia --color=yes --project=examples examples/hybrid/plane/density_current_2d_flux_form.jl"
+    artifact_paths:
+      - "examples/hybrid/output/dc_fluxform/*"
```
Comment (szy21) [inline .buildkite/pipeline.yml:405]:
```suggestion
      - "examples/hybrid/plane/output/dc_fluxform/*"
```

File: .buildkite/pipeline.yml [hunk 12]
```diff
@@ -374,6 +374,39 @@ steps:
...
+      config: cpu
+      queue: central
+      slurm_ntasks: 1
+  
+  - label: ":computer: Density current 2D hybrid invariant potential temperature"
+    key: "cpu_density_current_2d_hybrid_invariant_potential_temperature"
+    command:
+      - "julia --color=yes --project=examples examples/hybrid/plane/density_current_2dinvariant_rhotheta.jl"
+    artifact_paths:
+      - "examples/hybrid/output/dc_rhotheta/*"
```
Comment (szy21) [inline .buildkite/pipeline.yml:394]:
```suggestion
      - "examples/hybrid/plane/output/dc_rhotheta/*"
```

**Review by szy21 — CHANGES_REQUESTED** [review]
I just missed this - please change the artifact_paths so that the artifacts will show up in buildkite.

**Comment by akshaysridhar** [comment]
bors try

**Review by szy21 — APPROVED** [review]
